p2panda-rs/src/test_utils/db/stores/entry.rs

```diff
@@ -0,0 +1,355 @@
+    }
+
+    /// Returns the given range of log entries.
+    async fn get_paginated_log_entries(
+        &self,
+        author: &Author,
+        log_id: &LogId,
+        seq_num: &SeqNum,
+        max_number_of_entries: usize,
+    ) -> Result<Vec<StorageEntry>, EntryStorageError> {
+        let mut entries: Vec<StorageEntry> = Vec::new();
+        let mut seq_num = *seq_num;
+
+        while entries.len() < max_number_of_entries {
+            match self.get_entry_at_seq_num(author, log_id, &seq_num).await? {
+                Some(next_entry) => entries.push(next_entry),
+                None => break,
+            };
+            match seq_num.next() {
+                Some(next_seq_num) => seq_num = next_seq_num,
+                None => break,
+            };
+        }
+        Ok(entries)
+    }
+
+    /// Return vector of all entries of a given schema
+    async fn get_entries_by_schema(
+        &self,
+        schema: &SchemaId,
+    ) -> Result<Vec<StorageEntry>, EntryStorageError> {
+        let entries = self.entries.lock().unwrap();
+
+        let entries: Vec<StorageEntry> = entries
+            .iter()
+            .filter(|(_, entry)| entry.operation().schema() == *schema)
+            .map(|(_, entry)| entry.to_owned())
+            .collect();
+
+        Ok(entries)
+    }
+
+    async fn get_certificate_pool(
+        &self,
+        author: &Author,
+        log_id: &LogId,
+        initial_seq_num: &SeqNum,
+    ) -> Result<Vec<StorageEntry>, EntryStorageError> {
+        let seq_num = initial_seq_num.as_u64();
+        let cert_pool_seq_nums: Vec<SeqNum> = get_lipmaa_links_back_to(seq_num, 1)
+            .iter()
+            // Unwrapping as we know this is a valid sequence number
+            .map(|seq_num| SeqNum::new(*seq_num).unwrap())
+            .collect();
+        let mut cert_pool: Vec<StorageEntry> = Vec::new();
+
+        for seq_num in cert_pool_seq_nums {
+            let entry = match self.get_entry_at_seq_num(author, log_id, &seq_num).await? {
+                Some(entry) => Ok(entry),
+                None => Err(EntryStorageError::CertPoolEntryMissing(seq_num.as_u64())),
+            }?;
+            cert_pool.push(entry);
+        }
+
+        Ok(cert_pool)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use std::convert::TryFrom;
+
+    use rstest::rstest;
+
+    use crate::entry::{sign_and_encode, Entry, EntrySigned, LogId, SeqNum};
+    use crate::identity::{Author, KeyPair};
+    use crate::operation::OperationEncoded;
+    use crate::schema::SchemaId;
+    use crate::storage_provider::traits::test_utils::{test_db, TestStore};
+    use crate::storage_provider::traits::{AsStorageEntry, EntryStore};
+    use crate::test_utils::db::{SimplestStorageProvider, StorageEntry};
+    use crate::test_utils::fixtures::{
+        entry, entry_signed_encoded, key_pair, operation_encoded, random_key_pair, schema,
+    };
+
+    #[rstest]
+    #[async_std::test]
```

Comment:
When you embed this in aquadoggo you would end up having two runtimes in the tests (async_std and tokio), which is something you want to avoid.

Could we change this to `tokio-test`? Its an own crate dedicated to testing :+1:

Reply:
How do you suggest using `tokio-test`? It doesn't use macros in the same way as above. I was thinking to just swap to `tokio` and use `#[tokio::test]` (or whatever it is).

Reply:
Aha! Yeah, I just found that in a random example on the internet, haven't looked deeper, but maybe its not providing what we need.